acme/dns_challenge.go

```diff
@@ -62,48 +63,125 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
+	if err != nil {
+		return false, fmt.Errorf("Could not lookup nameservers: %v", err)
+	}
+
+	for _, ns := range nameservers {
+		r, err := dnsQuery(fqdn, dns.TypeTXT, net.JoinHostPort(ns, "53"))
+		if err != nil {
+			return false, err
+		}
+
+		if r.Rcode != dns.RcodeSuccess {
+			return false, fmt.Errorf("NS %s returned RCode %s", ns, dns.RcodeToString[r.Rcode])
+		}
+
+		var found bool
+		for _, rr := range r.Answer {
+			if _, ok := rr.(*dns.TXT); ok {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false, fmt.Errorf("NS %s did not return expected TXT record", ns)
+		}
+	}
+
+	return true, nil
+}
+
+// dnsQuery directly queries the given authoritative nameserver.
+func dnsQuery(fqdn string, rtype uint16, nameserver string) (in *dns.Msg, err error) {
 	m := new(dns.Msg)
-	m.SetQuestion(domain+".", dns.TypeSOA)
-	c := new(dns.Client)
-	in, _, err := c.Exchange(m, "google-public-dns-a.google.com:53")
+	m.SetQuestion(fqdn, rtype)
+	m.SetEdns0(4096, false)
+	m.RecursionDesired = false
+
+	in, err = dns.Exchange(m, nameserver)
+	if err == dns.ErrTruncated {
+		tcp := &dns.Client{Net: "tcp"}
+		in, _, err = tcp.Exchange(m, nameserver)
+	}
+
+	return
+}
+
+// lookupNameservers returns the authoritative nameservers for the given domain name.
+func lookupNameservers(fqdn string) ([]string, error) {
+	var referralNameservers []string
+
+	// We start recursion at the gTLD origin
+	// so we don't have to manage root hints
+	labels := dns.SplitDomainName(fqdn)
+	tld := labels[len(labels)-1]
+	nss, err := net.LookupNS(tld)
 	if err != nil {
-		return false
+		return nil, fmt.Errorf("Failed to resolve TLD %s %v", tld, err)
 	}
 
-	var authorativeNS string
-	for _, answ := range in.Answer {
-		soa := answ.(*dns.SOA)
-		authorativeNS = soa.Ns
+	for _, ns := range nss {
+		referralNameservers = append(referralNameservers, ns.Host)
 	}
 
-	fallbackCnt := 0
-	for fallbackCnt < preCheckDNSFallbackCount {
-		m.SetQuestion(fqdn, dns.TypeTXT)
-		in, _, err = c.Exchange(m, authorativeNS+":53")
-		if err != nil {
-			return false
+	// Follow the referrals until we hit the authoritative NS
+	for depth := 0; depth < recursionMaxDepth; depth++ {
+		var r *dns.Msg
+		var err error
+		
+		for _, ns := range referralNameservers {
+			r, err = dnsQuery(fqdn, dns.TypeNS, net.JoinHostPort(ns, "53"))
+			if err != nil {
+				continue
+			}
+			if r.Rcode == dns.RcodeSuccess {
+				break
+			}
+			if r.Rcode == dns.RcodeNameError {
```

Comment:
Consider:

    $ dig any internal-name.willglynn.com @8.8.8.8

    ; <<>> DiG 9.8.3-P1 <<>> any internal-name.willglynn.com @8.8.8.8
    ;; global options: +cmd
    ;; Got answer:
    ;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 39092
    ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0

    ;; QUESTION SECTION:
    ;internal-name.willglynn.com.	IN	ANY

    ;; AUTHORITY SECTION:
    willglynn.com.		1799	IN	SOA	ns1.willglynn.com. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 300

    ;; Query time: 52 msec
    ;; SERVER: 8.8.8.8#53(8.8.8.8)
    ;; WHEN: Sat Feb  6 23:16:44 2016
    ;; MSG SIZE  rcvd: 110

In this case, querying `internal-name.willglynn.com` fails with `NXDOMAIN`, causing `lookupNameservers()` to exit at this location each time it's called. This ultimately causes `lego` to spin forever, long after the desired `_acme-challenge.internal-name.willglynn.com IN TXT` record became globally resolvable from `ns1`..`ns4.willglynn.com`.

A logging statement confirms that `referralNameservers` contains `ns1`..`ns4.willglynn.com` at the moment this branch exits the function, which is the correct set of authoritative servers for that FQDN.